Optionally create missing Flags, Samples and Switches

adamhaney opened the pull request:
Flags, Switches and Samples may be scattered throughout templates,
backend code or across mulitple frontend clients. This patch provides
a user with the option of having waffle create a missing Flag, Sample
or Switch when one is requested but does not exist.

The goal of this change is to save operators time by removing the need
to search through code for flag/switch/sample names and to reduce
communication requirements accross teams with of multiple developers
with multiple environments that might require the creation of a
flag/switch/sample in the database of every environment code is run
in.

# Behavior Changes
## Old Behavior
  1. Flag/Switch/Sample is requested
  2a. If found in db evaulate what the value should be for request
  2b. If missing return (FLAG|SWITCH|SAMPLE)_DEFAULT value
## New Behavior
  1. Flag/Switch/Sample is requested
  2a. If found in db evaulate what the value should be for request
  2b. Create Flag/Switch/Sample with
      value (FLAG|SWITCH|SAMPLE)_DEFAULT
  3. Return (FLAG|SWITCH|SAMPLE)_DEFAULT for this request, subsequent
  requests will be served using the value from the cache/db

# Further Discussion
  - #44 (https://github.com/jsocol/django-waffle/issues/44)
  - #50 (https://github.com/jsocol/django-waffle/issues/50)

# Future Work
This change intentionally narrowly scoped and as such does not provide
a mechanism to set a default value other
than (FLAG|SWITCH|SAMPLE)_DEFAULT for individual objects that are not
in the database although it could be expanded in the future to do so.

adamhaney committed:
Optionally create missing Flags, Samples and Switches

Flags, Switches and Samples may be scattered throughout templates,
backend code or across mulitple frontend clients. This patch provides
a user with the option of having waffle create a missing Flag, Sample
or Switch when one is requested but does not exist.

The goal of this change is to save operators time by removing the need
to search through code for flag/switch/sample names and to reduce
communication requirements accross teams with of multiple developers
with multiple environments that might require the creation of a
flag/switch/sample in the database of every environment code is run
in.

Behavior Changes
================

Old Behavior
-----------
  1. Flag/Switch/Sample is requested
  2a. If found in db evaulate what the value should be for request
  2b. If missing return (FLAG|SWITCH|SAMPLE)_DEFAULT value

New Behavior
------------
  1. Flag/Switch/Sample is requested
  2a. If found in db evaulate what the value should be for request
  2b. Create Flag/Switch/Sample with
      value (FLAG|SWITCH|SAMPLE)_DEFAULT
  3. Return (FLAG|SWITCH|SAMPLE)_DEFAULT for this request, subsequent
  requests will be served using the value from the cache/db

Further Discussion
------------------
  - #44 (https://github.com/jsocol/django-waffle/issues/44)
  - #50 (https://github.com/jsocol/django-waffle/issues/50)

Future Work
-----------
This change intentionally narowly scoped and as such does not provide
a mechanism to set a default value other
than (FLAG|SWITCH|SAMPLE)_DEFAULT for individual objects that are not
in the database although it could be expanded in the future to do so.

adamhaney committed:
Optionally create missing Flags, Samples and Switches

Flags, Switches and Samples may be scattered throughout templates,
backend code or across mulitple frontend clients. This patch provides
a user with the option of having waffle create a missing Flag, Sample
or Switch when one is requested but does not exist.

The goal of this change is to save operators time by removing the need
to search through code for flag/switch/sample names and to reduce
communication requirements accross teams with of multiple developers
with multiple environments that might require the creation of a
flag/switch/sample in the database of every environment code is run
in.

  1. Flag/Switch/Sample is requested
  2a. If found in db evaulate what the value should be for request
  2b. If missing return (FLAG|SWITCH|SAMPLE)_DEFAULT value

  1. Flag/Switch/Sample is requested
  2a. If found in db evaulate what the value should be for request
  2b. Create Flag/Switch/Sample with
      value (FLAG|SWITCH|SAMPLE)_DEFAULT
  3. Return (FLAG|SWITCH|SAMPLE)_DEFAULT for this request, subsequent
  requests will be served using the value from the cache/db

  - #44 (https://github.com/jsocol/django-waffle/issues/44)
  - #50 (https://github.com/jsocol/django-waffle/issues/50)

This change intentionally narowly scoped and as such does not provide
a mechanism to set a default value other
than (FLAG|SWITCH|SAMPLE)_DEFAULT for individual objects that are not
in the database although it could be expanded in the future to do so.

adamhaney committed:
Merge branch 'master' of github.com:adamhaney/django-waffle

adamhaney committed:
Adding exception handling for IntegrityErrors on waffle creation

clintonb commented on waffle/models.py:
```diff
@@ -243,6 +243,19 @@ def _is_active_for_language(self, request):
+                            name=self.name,
+                            everyone=get_setting('FLAG_DEFAULT')
+                        )
+                except IntegrityError:
```
What would cause these errors? If such an error occurs, I would rather see it bubble up that be squashed with an uninformative log line.

adamhaney replied:
The IntegrityError would be thrown if you attempt to create a flag name that already exists, which can happen if two processes both encounter a flag that hasn't been created and attempt to create it. I caught it because I my philosophy around this change set was very "first do no harm" since it's intended to be a convenience and the behavior of returning the FLAG_DEFAULT is unaffected by whether or not the flag could be created I thought it was best to catch it.

Still, you're the maintainer so if you don't like that explanation I can definitely change it.

clintonb replied:
Log as a warning, but include the stack trace by passing `exc_info=True` in your call to `warning`. Do that in all of the new exception handlers

clintonb commented on waffle/models.py:
```diff
@@ -481,5 +508,23 @@ class Meta:
+                default_percent = 0.0
+
+                if get_setting('SAMPLE_DEFAULT'):
+                    default_percent = 100.0
```
You want to use the setting value. This can be combined with the line above: `get_setting('SAMPLE_DEFAULT', 0)`.

adamhaney replied:
I wasn't trying to just get a default value, or at least I don't think I was :)

In the defaults `SAMPLE_DEFAULT` is a boolean value, however there isn't a boolean option for a Sample, it's expressed as a percentage, so I'm saying if `SAMPLE_DEFAULT` is true then that implies that you want the sample value to be 100, if I just grabbed the default value I'm testing if it's defined.

I've never really used Samples in production so I may be misunderstanding their intended use.

clintonb replied:
Got it. I've never used samples, either.

clintonb commented on waffle/models.py:
```diff
@@ -243,6 +243,19 @@ def _is_active_for_language(self, request):
         if not self.pk:
+            if get_setting('CREATE_MISSING_FLAGS'):
+                try:
+                    with transaction.atomic():
```
Why are you creating a transaction for all of these?

adamhaney replied:
In Postgres (not sure about other databases) if a cursor throws an IntegrityError then later interactions with the database will fail (even if you catch the exception) [1]. So I'm just trying to make sure that failure to create flags doesn't take down what would otherwise be a perfectly healthy request.

1. https://docs.djangoproject.com/en/2.1/topics/db/transactions/#controlling-transactions-explicitly

Although, upon further thought I could probably accomplish a similar thing by using `get_or_create` instead of just `create`.

clintonb reviewed (changes requested):
I must admit that I am not a fan of implicitly creating data in this manner. You run the risk of creating a bunch of data that needs to later be cleaned up. Why not use a migration to create the flag/sample/switch?

adamhaney committed:
Tests for dynamic creation of Flags/Switches/Samples

These test to make sure that all models can be created dynamically
when the dynamic creation setting is active and that the dynamically
created object is instantiated appropriately based on the WAFFLE_*_DEFAULT
setting value

adamhaney committed:
Merge pull request #1 from adamhaney/feature/create-missing

Tests for dynamic creation of Flags/Switches/Samples

adamhaney committed:
Removing stray print statements from tests

adamhaney committed:
Merge branch 'master' of github.com:adamhaney/django-waffle into feature/create-missing

clintonb commented on waffle/models.py:
```diff
@@ -481,5 +508,23 @@ class Meta:
+            if get_setting('CREATE_MISSING_SAMPLES'):
+                default_percent = 0.0
+
+                if get_setting('SAMPLE_DEFAULT'):
```
`default_percent = 100 if get_setting('SAMPLE_DEFAULT') else 0`

clintonb commented on waffle/tests/test_waffle.py:
```diff
@@ -334,6 +334,35 @@ def test_pluggable_model(self):
+        assert not waffle.flag_is_active(get(), 'my_dynamically_created_flag')
+        assert flag_model.objects.count() == 1
+
+        flag = flag_model.objects.all()[0]
```
Combine into a single call: 
```python
FLAG_NAME = 'my_dynamically_created_flag'
flag = flag_model.objects.get(name=FLAG_NAME)```

clintonb commented on waffle/tests/test_waffle.py:
```diff
@@ -334,6 +334,35 @@ def test_pluggable_model(self):
+    @override_settings(WAFFLE_CREATE_MISSING_FLAGS=True)
+    @override_settings(WAFFLE_FLAG_DEFAULT=False)
+    def test_flag_created_dynamically_default_false(self):
+        flag_model = waffle.get_waffle_flag_model()
```
Extract all of this into a helper method (e.g. `assert_flag_dynamically_created(expected_value)`).

clintonb commented on waffle/tests/test_waffle.py:
```diff
@@ -396,6 +425,28 @@ def test_read_from_write_db(self):
+    @override_settings(WAFFLE_CREATE_MISSING_SWITCHES=True)
+    @override_settings(WAFFLE_SWITCH_DEFAULT=False)
+    def test_switch_created_dynamically_false(self):
+        assert Switch.objects.count() ==  0
```
Same notes as above: create a helper method, load the object by name.

clintonb commented on waffle/tests/test_waffle.py:
```diff
@@ -429,6 +480,31 @@ def test_read_from_write_db(self):
+    @override_settings(WAFFLE_CREATE_MISSING_SAMPLES=True)
+    @override_settings(WAFFLE_SAMPLE_DEFAULT=False)
+    def test_sample_created_dynamically_default_false(self):
+        assert Sample.objects.count() ==  0
```
Same notes as above: create a helper method, load the object by name.

adamhaney committed:
Instead of wrapping in transactions use get_or_create with defaults

adamhaney committed:
Merge pull request #2 from adamhaney/feature/create-missing

Instead of wrapping in transactions use get_or_create with defaults

adamhaney committed:
Refactoring test methods as requested in code review

adamhaney committed:
Documenting auto creation behavior for each type

clintonb commented on docs/starting/configuring.rst:
```diff
@@ -62,3 +62,19 @@ behavior.
     Which cache to use. Defaults to ``'default'``.
+
+``WAFFLE_CREATE_MISSING_FLAGS``
+    If waffle encounters a reference to a flag that is not in the database should it create it?
```
```suggestion
    If Waffle encounters a reference to a flag that is not in the database, should Waffle create the flag?
```

clintonb commented on docs/starting/configuring.rst:
```diff
@@ -62,3 +62,19 @@ behavior.
+    Defaults to ``False``.
+
+``WAFFLE_CREATE_MISSING_SWITCHES``
+    If waffle encounters a reference to a switch that is not in the database should it create it?
```
```suggestion
    If Waffle encounters a reference to a switch that is not in the database, should Waffle create the sample?
```

clintonb commented on docs/starting/configuring.rst:
```diff
@@ -62,3 +62,19 @@ behavior.
+    Defaults to ``False``.
+
+``WAFFLE_CREATE_MISSING_SAMPLES``
+    If waffle encounters a reference to a sample that is not in the database should it create it?
```
```suggestion
    If Waffle encounters a reference to a sample that is not in the database, should Waffle create the sample?
```

clintonb commented on docs/types/sample.rst:
```diff
@@ -52,3 +52,20 @@ Samples can be administered through the Django `admin site`_ or the
+``Percent`` attribute of the sample will be created as 100.0 (so that
+when the sample is checked it always evalues to 
+``True``). Otherwise the value will be set to 0.0 so that the sample
+always evaluates to ``False``.`
```
```suggestion
always evaluates to ``False``.
```

clintonb commented on docs/types/flag.rst:
```diff
@@ -207,3 +207,19 @@ visitors" to "percent chance that the Flag will be activated per visit."
+<starting-configuring>` to ``True``. Missing flags will be created in
+the database and the value of the ``Everyone`` flag attribute will be
+set to :ref:`WAFFLE_FLAG_DEFAULT <starting-configuring>` in the
+auto created database record.
```
```suggestion
auto-created database record.
```

clintonb commented on docs/types/switch.rst:
```diff
@@ -24,3 +24,18 @@ Switches can be administered through the Django `admin site`_ or the
+switch returns the :ref:`WAFFLE_SWITCH_DEFAULT <starting-configuring>`
+value but does not create a switch in the database. If you'd like waffle
+to create missing switchs in the database whenever it encounters a
+missing switch you can set :ref:`WAFFLE_CREATE_MISSING_SWITCHS
```
```suggestion
missing switch you can set :ref:`WAFFLE_CREATE_MISSING_SWITCHES
```

clintonb commented on docs/types/switch.rst:
```diff
@@ -24,3 +24,18 @@ Switches can be administered through the Django `admin site`_ or the
+value but does not create a switch in the database. If you'd like waffle
+to create missing switchs in the database whenever it encounters a
+missing switch you can set :ref:`WAFFLE_CREATE_MISSING_SWITCHS
+<starting-configuring>` to ``True``. Missing switchs will be created in
```
```suggestion
<starting-configuring>` to ``True``. Missing switches will be created in
```

clintonb commented on docs/types/switch.rst:
```diff
@@ -24,3 +24,18 @@ Switches can be administered through the Django `admin site`_ or the
+<starting-configuring>` to ``True``. Missing switchs will be created in
+the database and the value of the ``Active`` switch attribute will be
+set to :ref:`WAFFLE_SWITCH_DEFAULT <starting-configuring>` in the
+auto created database record.
```
```suggestion
auto-created database record.
```

clintonb committed:
documentation formatting fix

Co-Authored-By: adamhaney <adamhaney@users.noreply.github.com>

clintonb committed:
Fixing grammatical error in documentation

Co-Authored-By: adamhaney <adamhaney@users.noreply.github.com>

clintonb committed:
Fixing typo issue in documentation

Co-Authored-By: adamhaney <adamhaney@users.noreply.github.com>

clintonb committed:
Fixing formatting issue in documentation

Co-Authored-By: adamhaney <adamhaney@users.noreply.github.com>

clintonb committed:
Fixing syntax error in documentation formatting

Co-Authored-By: adamhaney <adamhaney@users.noreply.github.com>

clintonb committed:
Fixing capitalization issue in documentation

Co-Authored-By: adamhaney <adamhaney@users.noreply.github.com>

clintonb committed:
Fixing capitalization issue in documentation

Co-Authored-By: adamhaney <adamhaney@users.noreply.github.com>

clintonb committed:
Fix to capitalization issue in documentation

Co-Authored-By: adamhaney <adamhaney@users.noreply.github.com>
